sdk/core/System.ClientModel/gen/ModelReaderWriterContextGenerator.Emitter.cs

```diff
@@ -174,7 +177,20 @@ private void EmitContextClass(ModelReaderWriterContextGenerationSpec contextGene
             indent--;
             builder.AppendLine(indent, "}");
 
-            AddSource($"{contextName}.g.cs", SourceText.From(builder.ToString(), Encoding.UTF8));
+            AddNewFile(contextName, builder.ToString(), hintNames);
+        }
+
+        private void AddNewFile(string fileName, string source, HashSet<string> hintNames)
+        {
+            var hintName = fileName;
+            int collisionCount = 0;
+            while (hintNames.Contains(hintName))
+            {
+                hintName = $"{fileName}_{collisionCount}";
+            }
```

Comment:
Did you consider a `Dictionary<string, int>` here to avoid multiple lookups for collisions?